libvmi/os/windows/core.c

```diff
@@ -661,6 +661,280 @@ get_kpgd_from_rekall_profile(vmi_instance_t vmi)
+    }
+
+    uint32_t segment = (uint32_t)buffer[7] << 24
+                    | (uint32_t)buffer[4] << 16 
+                    | (uint32_t)buffer[3] << 8
+                    | buffer[2];
+
+    /* We have the starting address of the segment.
+    * Now we need to determine offset of Interrupt Service Routine
+    * inside of this segment, using IDT descriptor
+    * */
+
+    uint64_t address = 0;
+    page_mode_t mode = vmi_get_page_mode(vmi, 0);
+    switch(mode)
+    {
+        case VMI_PM_IA32E:
+            if (VMI_FAILURE == get_interrupt_routine_offset_64(vmi, idt_descriptor_address, &address))
+            {
+                return VMI_FAILURE;
+            }
+            break;
+        case VMI_PM_LEGACY:
+        case VMI_PM_PAE:
+            if (VMI_FAILURE == get_interrupt_routine_offset_32(vmi, idt_descriptor_address, &address)) 
+            {
+                return VMI_FAILURE;
+            }
+            break;
+        default:
+            return VMI_FAILURE;
+    };
+
+    *out_addr = address + segment;       // segment start + offset
+
+    return VMI_SUCCESS;
+}
+
+static bool
+nt_kernel_name_check(const char* name)
+{
+    return (0 == strncasecmp(name, "nt", 2) ||
+            0 == strncasecmp(name, "wrkx", 4) || 
+            0 == strncasecmp(name, "xnt", 3));
+}
+
+static status_t
+find_ntoskrnl_va(vmi_instance_t vmi, addr_t start_address_va, addr_t* ntoskrnl_address_va)
+{
+    status_t ret = VMI_FAILURE;
+    const size_t kernel_image_name_max_size = 8;
+    const uint64_t mask_4k = VMI_PS_4KB - 1;
+
+    uint8_t* buffer = malloc(VMI_PS_4KB);
+    if (NULL == buffer)
+    {
+        return ret;
+    }
+
+    struct export_table export_tbl;
+    addr_t et_rva = 0;
+    size_t et_size = 0;
+
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_PID,
+        .addr = 0,
+        .pid = 0
+    };
+
+    // virtual address of PE-image is always aligned to 4K boundaries
+    start_address_va &= ~mask_4k;
+
+    size_t bytes_readen = 0;
+    while (start_address_va > VMI_PS_4KB)       // do not check virtual address 0x1000, because there is no kernel there
+    {
+        // searching backwards for PE image
+
+        if (VMI_FAILURE != vmi_read_va(vmi, start_address_va, 0, VMI_PS_4KB, buffer, &bytes_readen) || 
+            bytes_readen != VMI_PS_4KB)
+        {
+            goto exit;
+        }
+
+        // check for PE-signature here
+        if (VMI_FAILURE == peparse_validate_pe_image(buffer, VMI_PS_4KB))
+        {
+            start_address_va -= VMI_PS_4KB;
+            continue;
+        }
+
+        // it is a PE-image. Let's get the module name
+        ctx.addr = start_address_va;
+        if (VMI_SUCCESS == peparse_get_export_table(vmi, &ctx, &export_tbl, &et_rva, &et_size))
+        {
+            et_rva = start_address_va + export_tbl.name;
+
+            if (kernel_image_name_max_size != vmi_read_va(vmi, et_rva, 0, kernel_image_name_max_size, buffer, &bytes_readen) || 
```

Comment:
This is incorrect, `vmi_read_va` returns a `status_t`.

Reply:
Sorry, my mistake. I was "porting" it from the older version, where this function returned number of bytes, written into the buffer.